### PR TITLE
std.string.indexOf_rangifed_fixup

### PR DESCRIPTION
**Pull request by burner**
make indexOf consistend and accept auto ref ranges

### CODE REVIEW

**Comment by burner** [comment]
@MartinNowak first blood again

File: std/string.d [hunk 1]
```diff
@@ -350,22 +350,24 @@ alias CaseSensitive = Flag!"caseSensitive";
...
-    if ( (isInputRange!Range && isSomeChar!(ElementEncodingType!Range)) ||
-        __traits(compiles, StringTypeOf!Range)
-    )
+    if (!(isInputRange!Range && isSomeChar!(ElementEncodingType!Range))
+        && is(StringTypeOf!Range))
+{
+    return indexOf(cast(StringTypeOf!Range)s, c, cs);
+}
+
+///
```
Comment (rainers) [inline std/string.d]: The documentation seems broken at this point. The order is changed and there are some parameter descriptions missing. Needs a "ditto"?

**Comment by burner** [comment]
@rainers looks better now

**Comment by rainers** [comment]
Yes, better, but the listed constraints are pretty confusing. I think we should hide the new overload by moving it below the original function and not document it at all (but maybe add a source code comment that it is used for alias this'd strings).

**Comment by burner** [comment]
Hm, but than people might think that they can not use the function, even though they could.
I think it is better to have a complex constraint, than to omit the functions from documentation.
I mean, what is the point of making a function callable but don't document its existence.

**Comment by rainers** [comment]
IMO constraints for overloads are an implementation detail to cover the desired range of types that are accepted by the template. Some of the constraints in the documentation are hardly decipherable, having a readable description of the intent instead is much more useful.

Especially in the case of aliased types, I rather think it is unexpected that the original implementation does not work (even to Walter who made most of the rangifications), so I doubt having an undocumented "StringTypeOf" in the constraint makes it clearer.

**Comment by rainers** [comment]
BTW: there is an undocumented (now removed) overload for fixed size arrays in the existing source, too.

If it is current phobos style to list all overloads, I'm ok with this PR. I would put the more complex version below the simpler one, though.

**Comment by burner** [comment]
I had a look at std.algorithm.comparision and at least there all overloads are in the documentation.
I think the undocumented(removed/hacky) overload was an error in the first place.

Therefore, I still think that all overloads should be shown(documented). After then I will consolidate all of phobos to do the same.

**Comment by rainers** [comment]
Ok, not topic of this PR to change phobos documentation style ;-)

**Comment by rainers** [comment]
Auto-merge toggled on

**Comment by burner** [comment]
> Ok, not topic of this PR to change phobos documentation style ;-) 

thanks

**Comment by burner** [comment]
@rainers this is wired, it passes the tests, auto-merge is on but It didn't merge.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by DmitryOlshansky** [comment]
Somehow it toggled back

**Comment by burner** [comment]
hm, strange. thanks anyway

**Pull request merged**